Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+ */
+
+
+template<size_t K, typename Data=su2double, typename Index=unsigned long>
+class NdFlattener;
+
+namespace helpers {
+  template<typename MPI_Allgatherv_type, typename MPI_Datatype_type, typename MPI_Communicator_type>
+  struct NdFlattener_MPI_Environment {
+    MPI_Allgatherv_type MPI_Allgatherv;
+    MPI_Datatype_type mpi_data;
+    MPI_Datatype_type mpi_index;
+    MPI_Communicator_type comm;
+    int rank; int size;
+  };
+
+  /*! \class IndexAccumulator
+   * \brief Data structure holding an offset for the NdFlattener, to provide a []...[]-interface.
+   * \details Derived from IndexAccumulator_Base, specifying the operator[] method:
+   *  - For N==1, the structure has already read all indices but one. So after this method has read the last
+   *    index, return a reference to the data.
+   *  - The case N==2 is much like the case N>3 but an additional function data() should be provided.
+   *  - For N>3, more indices have to be read, return an IndexAccumulator<N-1>.
+   * \tparam N - Number of missing parameters
+   * \tparam Nd_type - Type of the accessed NdFlattener, should be NdFlattener<N,...>
+   */
+  /*! \class IndexAccumulator_Base
+   * \brief Parent class of IndexAccumulator.
+   * \details IndexAccumulator provides the operator[] method.
+   */
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator_Base {
+  public:
+    static constexpr size_t N = _N;
+    using Nd_type = _Nd_type;
+
+    /*== Conditional is part of the standard library since C++14, keep it here for C++11 compatibility. ==*/
+    template<bool B, class T, class F> struct conditional { typedef T type; };
+    template<class T, class F> struct conditional<false, T, F> { typedef F type; };
+
+  protected:
+    Nd_type* nd; /*!< \brief The accessed NdFlattener. */
+    const typename Nd_type::Index offset; /*!< \brief Index in the currently accessed layer. */
+    //const typename Nd_type::Index size; /*!< \brief Exclusive upper bound for the next index. */
+    IndexAccumulator_Base(Nd_type* nd, typename Nd_type::Index offset):
+      nd(nd), offset(offset) {}
+
+  };
+
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator : public IndexAccumulator_Base<_N,_Nd_type>{
+  public:
+    static constexpr size_t N = _N;
+    using Nd_type = _Nd_type;
+    typedef IndexAccumulator_Base<_N,_Nd_type> Base;
+
+    template<class ...ARGS> IndexAccumulator(ARGS... args): Base(args...) {}
+
+    /*! The Base of NdFlattener<K> is NdFlattener<K-1>, but do also preserve constness.
+     */
+    using Nd_Base_type = typename Base::template conditional<
+      std::is_const<Nd_type>::value,
+      const typename Nd_type::Base,
+      typename Nd_type::Base
+    >::type;
+    /*! Return type of operator[]. */
+    using LookupType = IndexAccumulator<N-1, Nd_Base_type>;
+
+    /*! \brief Read one more index.
+     * \param[in] i - Index.
+     */
+    LookupType operator[] (typename Nd_type::Index i) {
```

Comment:
Actually, it was possible to make `IndexAccumulator<N,...>::operator[]` and also `IndexAccumulator<1,...>::data` always const methods. The accumulator just knows how to access an `NdFlattener` (and which one), and this information does not change along with the access.  
The exact type of the `NdFlattener`, including a `const`-qualifier, is a template argument to `IndexAccumulator`. This constness is preserved by `operator[]` and eventually used to determine whether the reference/pointer to the data should be const.

Reply:
I get that but if [ ] is const this should const-qualify the nd& kept by base which should then only be able to return const&... :thinking:
Have you tried modifying data via the IndexAccumulator?

Reply:
const method qualifies "this" but not the member, that is why it works...
However, this would allow a const IndexAccumulator to be changed https://gcc.godbolt.org/z/chvfrhEc6.

Reply:
This is how I understand what is happening:

In a const method `X::f()`, `this` is not `X*` but `X const*`, making any member read-only.

Anyway, a reference member, like `IndexAccumulator<N,Nd_t>::nd` of type `Nd_t&`, is always read-only in the sense that it cannot be reassigned to point to another variable. Whether or not the variable to which it points can be changed, does not depend on the constness of `*this`. It depends only on whether the type `Nd_t` is const, which can be found out with `std::is_const<Nd_t>::value`.

If the `IndexAccumulator` was generated by the `operator[]` from a non-const `NdFlattener`, it's an (non-const) `IndexAccumulator<N, NdFlattener<...> >` so its member `nd` is a reference to a non-const `NdFlattener`, so after supplying enough indices we get a reference/pointer to the `IndexAccumulator<1, NdFlattener<...> >::LookupType` aka `Data_t`.

If the `IndexAccumulator` was generated by the `operator[] const` from a const `NdFlattener`, it's an (non-const) `IndexAccumulator<N, const NdFlattener<...> >` so its member `nd` is a reference to a `const NdFlattener`, so after supplying enough indices we get a reference/pointer to the `IndexAccumulator<1, const NdFlattener<...> >::LookupType` aka `const Data_t`.

To simulate this in the godbolt example, replacing `vector<int>` by `const vector<int>` raises a compilation error "assignment of read-only location". In `ndflattener_tests.cpp`, I tried writing to `nd3_const[0][0][0]` and it did not compile. I don't know whether we can make a unit test for "does not compile".